### ISSUE TITLE
Junit5 - Run all callback methods on the class hierarchy

### ISSUE THREAD

**Issue by denniskawurek** [database-rider/database-rider#263]
Hi everyone,

I have the following basic test class, which wipes the data of the H2-database and leaves the *basic data* needed for all tests:

```
@ExtendWith(DBUnitExtension.class)
public class BaseTest {
    @BeforeEach
    @DataSet(cleanBefore = true, skipCleaningFor = {
        "client"
    })
    public void tearDownDB() {}
}
```

Then I have this test class

```
@SpringBootTest
public class ConcreteTest extends BaseTest {
    @BeforeEach
    @DataSet(executeScriptsAfter = "sql/test-setup.sql")
    public void init() {
    }

    @Test
    public void testCase1() {
    }

    @Test
    public void testCase2() {
    }
}
```

I tried to make it run with two different ways:

1. `executeScriptsAfter` in `ConcreteTest`

First the `@BeforeEach` method of the `BaseTest` is called and afterwards the one of the `ConcreteTest`.

In this case there is no database error, but my tests fail because the data doesn't seem to be populated to the database?

Beside of that I get this warning:

```
2021-03-12 14:06:22.352  WARN 18228 --- [    Test worker] c.g.d.r.c.dataset.DataSetExecutorImpl    : Database will not be populated because no dataset has been provided.
```

2. `executeScriptsBefore` in `ConcreteTest` 
In this case I do get a `Foreign key` exception in the first and second test case, which let it look like the method is called twice?

Beside of that, annotating every `@Test` method with `@DataSet(executeScriptsAfter = "sql/test-setup.sql")` works perfect!

I want to ask if the above test case is possible with database-rider or how I may help to debug it or so?
I am using `rider-junit5:1.23.0`

Thanks in advance.

**Comment by rmpestano** [comment]
Hi @denniskawurek, we have a test case for that, can you compare with your code? 

See example here: https://github.com/database-rider/database-rider/blob/master/rider-junit5/src/test/java/com/github/database/rider/junit5/JUnit5LifecycleHooksInSuperclassIt.java

**Comment by rmpestano** [comment]
When I find some time I'll try to add the same scenario to our [SpringBoot sample](https://github.com/database-rider/database-rider/blob/master/rider-examples/spring-boot-dbunit-sample/src/test/java/com/github/database/rider/springboot/SpringBootDBUnitTest.java), maybe it's related to SpringBootTest.

**Comment by denniskawurek** [comment]
Hi @rmpestano 
Thank you for your answer.

I've checked the code of the samples with my code and found some differences, but unfortunately changing it accordingly to the samples didn't work. I've got the same issues as above - either the data isn't stored or when not setting `disableConstraints = true` I'm getting an Primary key exception by the database as the data didn't get wiped.

I have changed my test case to the following. Please note that my H2 connection is set in the `application.yml` to:

```
@DBUnit
public abstract class BaseTest {
    @BeforeEach
    @DataSet(cleanBefore = true, skipCleaningFor = {
        "client"
    })
    public void tearDownDB() {}
}
```

Extending test class:

```
@SpringBootTest
@DBRider
public class ConcreteTest extends BaseTest {
    @BeforeEach
    @DataSet(executeScriptsAfter = "sql/test-setup.sql", disableConstraints = true)
    public void init() {
    }

    @Test
    public void testCase1() {
    }

    @Test
    public void testCase2() {
    }
}
```


```
    url: jdbc:h2:mem:db;DB_CLOSE_DELAY=-1;MODE=MySQL
    username: sa
    password: sa
    driver-class-name: org.h2.Driver
```

**Comment by cselagea** [comment]
I have the same issue.

Superclass:
```
@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@ContextConfiguration(initializers = DataSourcePropertiesInitializer.class)
@TestInstance(Lifecycle.PER_CLASS)
@DBRider
public abstract class BaseContractTest {

    @LocalServerPort
    private int port;

    @BeforeEach
    @DataSet("datasets/security.yml")
    public void setUp() {
        RestAssured.baseURI = "http://localhost";
        RestAssured.port = this.port;
    }

}
```

Subclass:
```
public abstract class BaseBookmarkTest extends BaseContractTest {

    @BeforeEach
    @DataSet(value = "datasets/bookmarks.yml", fillIdentityColumns = true)
    public void addBookmarks() {
    }

}
```

When I run a test class that extends `BaseBookmarkTest`, it fails because the security.yml dataset was not created.

I added some `println` statements to both of these `@BeforeEach` methods and observed that both methods are being executed, but `com.github.database.rider.junit5.DBUnitExtension#beforeEach` is only called once and it uses `findFirst`, so it only picks one of the `@BeforeEach` methods:

```
private Optional<Method> findCallbackMethod(Class testClass, Class callback) {
    return Stream.of(testClass
            .getMethods())
            .filter(m -> m.getAnnotation(callback) != null)
            .findFirst();
}
```

If I add security.yml to `BaseBookmarkTest`, then my test passes.

@rmpestano I checked the sample test that you mentioned, but it does not cover the scenario in which there are multiple `@BeforeEach` methods with `@DataSet`.

So am I just using Database Rider incorrectly and I should only have `@DataSet` declared on one `@BeforeEach`, or can it be enhanced to create datasets declared in every `@BeforeEach` method in the class hierarchy?

**Comment by rmpestano** [comment]
Hey guys, so the idea is to run all callback methods in the class hierarchy. Looking the the code it seems to be possible by changing `findCallbackMethod` to return a list of methods and for each of them call [executeDataSet here](https://github.com/database-rider/database-rider/blob/52d14ea8436f0fa01a5d8cc844bf1437726e7001/rider-junit5/src/main/java/com/github/database/rider/junit5/DBUnitExtension.java#L315-L319)

Then add a BeforeEach on [this test ](https://github.com/database-rider/database-rider/blob/master/rider-junit5/src/test/java/com/github/database/rider/junit5/JUnit5LifecycleHooksInSuperclassIt.java#L19) to verify the behaviour.

Who wants to jump in and contribute? 

Thank you for the feedback!

**Comment by rmpestano** [comment]
Hi @denniskawurek  and @cselagea, can you try with 1.23.2-SNAPSHOT to check if the issue is fixed? Check here for how to enable snapshots: https://github.com/database-rider/database-rider#snapshots

**Comment by rmpestano** [comment]
The fix is available in v`1.24.0`, you can reopen this issue later if you still have issues with it

**Comment by cselagea** [comment]
Hey @rmpestano I finally got a chance to try this. It's working, but it doesn't seem to conform to the usual execution order for JUnit lifecycle methods, e.g. "`@BeforeAll` methods from superclasses will be executed before `@BeforeAll` methods in subclasses".

I have a test in which the superclass `@BeforeAll` is running after the `@BeforeAll` from the test class.
